numpyro/distributions/distribution.py

```diff
@@ -563,6 +564,38 @@ def is_discrete(self):
         return self.support.is_discrete
 
 
+@runtime_checkable
+class DistributionLike(Protocol):
```

Comment:
why Protocol is needed?

Reply:
This protocol is needed to provide static type checking for functions that accept
    distribution objects. From what I see, NumPyro supports multiple distribution implementations:
    
    - numpyro.distributions.Distribution
    - funsor.Funsor 
    - tensorflow_probability.distributions.Distribution
    
    Rather than checking for each specific type, this protocol defines the common interface
    that all distribution objects must implement. This allows static type checkers to verify
    that any distribution object passed to a function has the required methods and properties,
    regardless of which implementation it uses.
    
    If we remove it then we will get many MyPy errors 😅 .
    
    I have seen this approach in other projects so this is why I suggest it :)

Reply:
(And tensorflow probability and fusor are optional dependencies so we can not use Union to create a common type 🙃)

Reply:
SGTM! I was a bit worried of introducing a new abstraction but I trust the tests :)